mesonbuild/options.py

```diff
@@ -682,96 +704,370 @@ def add_to_argparse(self, name: str, parser: argparse.ArgumentParser, help_suffi
+        self.options[optkey].set_value(opt)
+
+    def set_option(self, key: OptionKey, new_value: str, first_invocation: bool = False):
+        assert isinstance(key, OptionKey)
+        # FIXME, dupe of set_value
+        # Remove one of the two before merging to master.
+        if key.name == 'prefix':
+            new_value = self.sanitize_prefix(new_value)
+        elif self.is_builtin_option(key):
+            prefix = self.get_value_for('prefix')
+            new_value = self.sanitize_dir_option_value(prefix, key, new_value)
+        opt = self.get_value_object_for(key)
+        if opt.deprecated is True:
+            mlog.deprecation(f'Option {key.name!r} is deprecated')
+        elif isinstance(opt.deprecated, list):
+            for v in opt.listify(new_value):
+                if v in opt.deprecated:
+                    mlog.deprecation(f'Option {key.name!r} value {v!r} is deprecated')
+        elif isinstance(opt.deprecated, dict):
+            def replace(v):
+                newvalue = opt.deprecated.get(v)
+                if newvalue is not None:
+                    mlog.deprecation(f'Option {key.name!r} value {v!r} is replaced by {newvalue!r}')
+                    return newvalue
+                return v
+            valarr = [replace(v) for v in opt.listify(new_value)]
+            new_value = ','.join(valarr)
+        elif isinstance(opt.deprecated, str):
+            mlog.deprecation(f'Option {key.name!r} is replaced by {opt.deprecated!r}')
+            # Change both this aption and the new one pointed to.
+            dirty = self.set_option(key.evolve(name=opt.deprecated), new_value)
+            dirty |= opt.set_value(new_value)
+            return dirty
+
+        old_value = opt.value
+        changed = opt.set_value(new_value)
+
+        if opt.readonly and changed:
+            raise MesonException(f'Tried modify read only option {str(key)!r}')
+
+        if key.name == 'prefix' and first_invocation and changed:
+            self.reset_prefixed_options(old_value, new_value)
+
+        if changed:
+            self.set_dependents(key, new_value)
+
+        return changed
+
+    def set_option_from_string(self, keystr, new_value):
+        if isinstance(keystr, OptionKey):
+            o = keystr
+        else:
+            o = OptionKey.from_string(keystr)
+        if o in self.options:
+            return self.set_value(o, new_value)
+        o = o.evolve(subproject='')
+        return self.set_value(o, new_value)
+
+    def set_subproject_options(self, subproject, spcall_default_options, project_default_options):
+        for o in itertools.chain(spcall_default_options, project_default_options):
+            keystr, valstr = o.split('=', 1)
+            assert ':' not in keystr
+            keystr = f'{subproject}:{keystr}'
+            if keystr not in self.augments:
+                self.augments[keystr] = valstr
+
+    def set_from_configure_command(self, D: T.List[str], U: T.List[str]) -> bool:
```

Comment:
the `D` and `U` here are not descriptive names, and they look like they should be generic values.